edb/edgeql/tokenizer.py

```diff
@@ -89,7 +89,12 @@ def extra_blob(self) -> bytes:
 
     @classmethod
     def from_string(cls, text: str) -> NormalizedSource:
-        return cls(normalize(text))
+        return cls(normalize(text), text)
```

Comment:
Perhaps make `cache_key` a member of `Source` instead and then you won't have to use a tuple?

Reply:
+1. I actually didn't like this method too.

Reply:
It though about this. But we already transfer query to the complier two times (once as a source and other time as a token list). Adding a cache_key here would add a third time. I remember we discussed few megabyte-sized queries, so that amount of tranfer may be significant. But cache_key is useless in the compiler.

Reply:
> I remember we discussed few megabyte-sized queries

Usually that was about SQL queries we generate. I'm not sure how long it would take to compile a megabyte of EdgeQL, but that time will likely dwarf the socket io time.

That said, I agree, transferring cache_key to compilers does seem wasteful.

Reply:
Maybe make `cache_key` a hash of source then?

Reply:
Almost lost this comment. Yes making it hash makes sense. I don't remember why it wasn't done this way from the start. Perhaps for easier debugging (i.e. `cache_key` contains text with variables extracted but in textual form, not as a list of tokens). I'll try to export something else for debugging and use a hash.

Reply:
Done